File: travel_deals/sources/serpapi_google_hotels.py
from __future__ import annotations

import json
import os
from collections.abc import Iterable
from urllib.parse import urlencode
from urllib.request import urlopen

from travel_deals.models import CityConfig, DateWindow, Evidence, Listing
from travel_deals.sources.base import SourceAdapter
from travel_deals.transit import estimate_transit_accessibility


class SerpApiGoogleHotelsSource(SourceAdapter):
    name = "serpapi_google_hotels"

    def __init__(self, api_key: str | None = None, timeout: int = 20):
        self.api_key = api_key or os.getenv("SERPAPI_API_KEY")
        self.timeout = timeout

    def search(self, cities: Iterable[CityConfig], windows: Iterable[DateWindow]) -> list[Listing]:
        if not self.api_key:
            return []

        listings: list[Listing] = []
        for city in cities:
            for window in windows:
                payload = self._fetch(city, window)
                for item in payload.get("properties", [])[:15]:
                    listing = self._normalize(item, city, window)
                    if listing is not None:
                        listing.transit = estimate_transit_accessibility(listing, city)
                        listings.append(listing)
        return listings

    def _fetch(self, city: CityConfig, window: DateWindow) -> dict:
        params = {
            "engine": "google_hotels",
            "q": f"aparthotel apartment hotel {city.name}",
            "check_in_date": window.check_in.isoformat(),
            "check_out_date": window.check_out.isoformat(),
            "adults": "1",
            "currency": "EUR",
            "gl": city.country.lower() or "us",
            "hl": "en",
            "api_key": self.api_key,
        }
        url = f"https://serpapi.com/search.json?{urlencode(params)}"
        with urlopen(url, timeout=self.timeout) as response:
            return json.loads(response.read().decode("utf-8"))

    def _normalize(self, item: dict, city: CityConfig, window: DateWindow) -> Listing | None:
        rate = item.get("rate_per_night") or item.get("total_rate")
        extracted = None
        if isinstance(rate, dict):
            extracted = rate.get("extracted_lowest") or rate.get("extracted_before_taxes_fees")
        elif isinstance(rate, (int, float)):
            extracted = float(rate)
        if not extracted:
            return None

        total = item.get("total_rate", {})
        total_price = total.get("extracted_lowest") if isinstance(total, dict) else None
        fees_unclear = total_price is None
        nightly = float(extracted)
        if total_price:
            total_price = float(total_price)
            nightly = total_price / window.nights
        else:
            total_price = nightly * window.nights

        gps = item.get("gps_coordinates") or {}
        amenities_text = " ".join(item.get("amenities", [])).lower()
        kitchen_status = "confirmed" if "kitchen" in amenities_text or "kitchenette" in amenities_text else "missing"
        air_status = "ambiguous"
        if "air conditioning" in amenities_text or "climate" in amenities_text:
            air_status = "ambiguous"

        amenities = {
            "adjustable_climate_control": Evidence(
                air_status,
                "Google Hotels may show air conditioning, but individual adjustability usually needs manual confirmation.",
                self.name,
            ),
            "kitchen_or_kitchenette": Evidence(kitchen_status, "Amenity text was checked for kitchen/kitchenette.", self.name),
            "stovetop": Evidence("ambiguous", "Search result amenities rarely distinguish a stovetop.", self.name),
            "utensils": Evidence("ambiguous", "Kitchen inventory is not reliably exposed by Google Hotels search data.", self.name),
            "blackout_window_covering": Evidence("ambiguous", "Blackout curtains/shutters are rarely exposed by Google Hotels search data.", self.name),
        }

        return Listing(
            source=self.name,
            source_reliability=0.72,
            source_listing_id=str(item.get("property_token") or item.get("name")),
            name=item.get("name", "Unnamed property"),
            url=item.get("link") or item.get("serpapi_property_details_link") or "",
            city=city.name,
            neighborhood=item.get("neighborhood", ""),
            dates=window,
            total_price_eur=round(total_price, 2),
            nightly_price_eur=round(nightly, 2),
            rating=item.get("overall_rating"),
            rating_scale=5.0,
            review_count=item.get("reviews"),
            location_text=item.get("location_rating", "") or item.get("description", ""),
            latitude=gps.get("latitude"),
            longitude=gps.get("longitude"),
            room_type=item.get("type", ""),
            cancellation_policy="",
            amenities=amenities,
            raw={**item, "fees_unclear": fees_unclear},
        )

File: src/sources/serpapiGoogleHotels.ts
import { randomUUID } from "node:crypto";
import type { CityConfig, DateWindow, Evidence, Listing } from "../models.js";
import { estimateTransitAccessibility } from "../transit.js";
import type { SourceAdapter } from "./base.js";

interface SerpApiHotelItem {
  name?: string;
  property_token?: string;
  link?: string;
  serpapi_property_details_link?: string;
  neighborhood?: string;
  type?: string;
  description?: string;
  location_rating?: string;
  overall_rating?: number;
  reviews?: number;
  amenities?: string[];
  gps_coordinates?: {
    latitude?: number;
    longitude?: number;
  };
  rate_per_night?: RateValue;
  total_rate?: RateValue;
  [key: string]: unknown;
}

type RateValue =
  | number
  | {
      extracted_lowest?: number;
      extracted_before_taxes_fees?: number;
      [key: string]: unknown;
    };

interface SerpApiResponse {
  properties?: SerpApiHotelItem[];
  error?: string;
}

export class SerpApiGoogleHotelsSource implements SourceAdapter {
  readonly name = "serpapi_google_hotels";

  constructor(
    private readonly apiKey = process.env.SERPAPI_API_KEY,
    private readonly maxPerSearch = 15
  ) {}

  async search(cities: CityConfig[], windows: DateWindow[]): Promise<Listing[]> {
    if (!this.apiKey) return [];

    const listings: Listing[] = [];
    for (const city of cities) {
      for (const window of windows) {
        const payload = await this.fetch(city, window);
        for (const item of (payload.properties ?? []).slice(0, this.maxPerSearch)) {
          const listing = this.normalize(item, city, window);
          if (listing) {
            listing.transit = estimateTransitAccessibility(listing, city);
            listings.push(listing);
          }
        }
      }
    }
    return listings;
  }

  private async fetch(city: CityConfig, window: DateWindow): Promise<SerpApiResponse> {
    const params = new URLSearchParams({
      engine: "google_hotels",
      q: `aparthotel apartment hotel ${city.name}`,
      check_in_date: window.checkIn,
      check_out_date: window.checkOut,
      adults: "1",
      currency: "EUR",
      gl: city.country.toLowerCase() || "us",
      hl: "en",
      api_key: this.apiKey ?? ""
    });
    const response = await fetch(`https://serpapi.com/search.json?${params.toString()}`);
    if (!response.ok) {
      throw new Error(`SerpAPI request failed: ${response.status} ${response.statusText}`);
    }
    return (await response.json()) as SerpApiResponse;
  }

  private normalize(item: SerpApiHotelItem, city: CityConfig, window: DateWindow): Listing | null {
    const rate = item.rate_per_night ?? item.total_rate;
    const extracted = extractRate(rate);
    if (!extracted) return null;

    const totalRate = extractRate(item.total_rate);
    const feesUnclear = totalRate === null;
    const totalPrice = totalRate ?? extracted * window.nights;
    const nightly = totalPrice / window.nights;
    const amenitiesText = (item.amenities ?? []).join(" ").toLowerCase();
    const kitchenStatus: Evidence["status"] =
      amenitiesText.includes("kitchen") || amenitiesText.includes("kitchenette") ? "confirmed" : "missing";
    const airStatus: Evidence["status"] =
      amenitiesText.includes("air conditioning") || amenitiesText.includes("climate") ? "ambiguous" : "missing";
    const amenities: Record<string, Evidence> = {
      adjustable_climate_control: {
        status: airStatus,
        detail: "Google Hotels may show air conditioning, but individual adjustability usually needs manual confirmation.",
        source: this.name
      },
      kitchen_or_kitchenette: {
        status: kitchenStatus,
        detail: "Amenity text was checked for kitchen/kitchenette.",
        source: this.name
      },
      stovetop: {
        status: "ambiguous",
        detail: "Search result amenities rarely distinguish a stovetop.",
        source: this.name
      },
      utensils: {
        status: "ambiguous",
        detail: "Kitchen inventory is not reliably exposed by Google Hotels search data.",
        source: this.name
      },
      blackout_window_covering: {
        status: "ambiguous",
        detail: "Blackout curtains/shutters are rarely exposed by Google Hotels search data.",
        source: this.name
      }
    };

    return {
      source: this.name,
      sourceReliability: 0.72,
      sourceListingId: String(item.property_token ?? item.name ?? randomUUID()),
      name: item.name ?? "Unnamed property",
      url: item.link ?? item.serpapi_property_details_link ?? "",
      city: city.name,
      neighborhood: item.neighborhood ?? "",
      dates: window,
      totalPriceEur: round(totalPrice, 2),
      nightlyPriceEur: round(nightly, 2),
      rating: item.overall_rating ?? null,
      ratingScale: 5,
      reviewCount: item.reviews ?? null,
      locationText: item.location_rating ?? item.description ?? "",
      latitude: item.gps_coordinates?.latitude ?? null,
      longitude: item.gps_coordinates?.longitude ?? null,
      roomType: item.type ?? "",
      cancellationPolicy: "",
      amenities,
      transit: { status: "missing", detail: "No transit evidence." },
      raw: { ...item, fees_unclear: feesUnclear }
    };
  }
}

function extractRate(rate: RateValue | undefined): number | null {
  if (typeof rate === "number") return rate;
  if (!rate) return null;
  return rate.extracted_lowest ?? rate.extracted_before_taxes_fees ?? null;
}

function round(value: number, digits: number): number {
  return Number(value.toFixed(digits));
}
